pycbc/inference/sampler/nessai.py

```diff
@@ -0,0 +1,332 @@
+
+    def finalize(self):
+        """Finalize sampling"""
+        logz = self._sampler.log_evidence
+        dlogz = self._sampler.log_evidence_error
+
+        logging.info(f"log Z, dlog Z: {logz}, {dlogz}")
+
+        self.checkpoint()
+
+        for fn in [self.checkpoint_file, self.backup_file]:
+            self.write_results(fn)
+
+    def write_results(self, filename):
+        """Write the results to a given file"""
+        with self.io(filename, "a") as fp:
+            fp.write_raw_samples(self.samples)
+            fp.write_logevidence(
+                self._sampler.log_evidence,
+                self._sampler.log_evidence_error,
+            )
+
+
+class NessaiModel(nessai.model.Model):
+    """Wrapper for PyCBC Inference model class for use with nessai.
+
+    Parameters
+    ----------
+    model : inference.BaseModel instance
+        A model instance from PyCBC.
+    loglikelihood_function : str
+        Name of the log-likelihood method to call.
+    """
+
+    def __init__(self, model, loglikelihood_function=None):
+        self.model = model
+        self.names = list(model.sampling_params)
+
+        # Configure the log-likelihood function
+        if loglikelihood_function is None:
+            loglikelihood_function = "loglikelihood"
+        self.loglikelihood_function = loglikelihood_function
+
+        # Configure the priors bounds
+        bounds = {}
+        for dist in model.prior_distribution.distributions:
+            bounds.update(
+                **{
+                    k: [v.min, v.max]
+                    for k, v in dist.bounds.items()
+                    if k in self.names
+                }
+            )
+        self.bounds = bounds
+        # Prior and likelihood are not vectorised
+        self.vectorised_likelihood = False
+        self.vectorised_prior = False
+        # Use the pool for computing the prior
+        self.parallelise_prior = True
+
+    def to_dict(self, x):
+        return {n: x[n].item() for n in self.names}
+
+    def to_live_points(self, x):
+        """Convert to the structured arrays used by nessai"""
+        # TODO: could this be improved?
+        return nessai.livepoint.numpy_array_to_live_points(
+            rfn.structured_to_unstructured(x),
+            self.names,
+        )
+
+    def new_point(self, N=1):
+        """Draw a new point"""
+        return self.to_live_points(self.model.prior_rvs(size=N))
+
+    def new_point_log_prob(self, x):
+        """Log-probability for the ``new_point`` method"""
+        return self.batch_evaluate_log_prior(x)
+
+    def log_prior(self, x):
+        """Compute the log-prior"""
+        self.model.update(**self.to_dict(x))
+        return self.model.logprior
+
+    def log_likelihood(self, x):
+        """Compute the log-likelihood"""
+        self.model.update(**self.to_dict(x))
+        return getattr(self.model, self.loglikelihood_function)
+
+    def from_unit_hypercube(self, x):
```

Comment:
This direction can be implemented pretty straightforwardly, see https://github.com/gwastro/pycbc/blob/master/pycbc/inference/sampler/base_cube.py#L82

Reply:
Thanks, I'll look at adding that. Unfortunately, the current implementation requires both directions, though it's something I'm hoping to change in a future release.

Reply:
Ok, I don't think that's unreasonable, but would require some changes to the distribution package to support.